tmt/utils.py

```diff
@@ -2632,21 +2634,44 @@ def remove_color(text: str) -> str:
     return re.sub(r'\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])', '', text)
 
 
-def default_branch(repository: Path, remote: str = 'origin') -> Optional[str]:
+def default_branch(repository: Path, logger: tmt.log.Logger,
+                   remote: str = 'origin') -> Optional[str]:
     """ Detect default branch from given local git repository """
-    # Make sure the '.git/refs/remotes/{remote}' directory is present
-    git_remotes_dir = repository / f'.git/refs/remotes/{remote}'
-    if not git_remotes_dir.exists():
+    # Make sure '.git' (which is a file or a directory) is present
+    dot_git = repository / '.git'
+    if not dot_git.exists():
+        return None
+
+    if not dot_git.is_file() and not dot_git.is_dir():
         return None
 
-    head = repository / f'.git/refs/remotes/{remote}/HEAD'
-    # Make sure the HEAD reference is available
-    if not head.exists():
-        subprocess.run(
-            f'git remote set-head {remote} --auto'.split(), cwd=repository)
-    # The ref format is 'ref: refs/remotes/origin/main'
-    with open(head) as ref:
-        return ref.read().strip().split('/')[-1]
+    if dot_git.is_dir():
+        # Make sure the '.git/refs/remotes/{remote}' directory is present
+        git_remotes_dir = repository / f'.git/refs/remotes/{remote}'
+        if not git_remotes_dir.exists():
+            return None
+
+        head = repository / f'.git/refs/remotes/{remote}/HEAD'
+        # Make sure the HEAD reference is available
+        if not head.exists():
+            subprocess.run(
+                f'git remote set-head {remote} --auto'.split(), cwd=repository)
+        # The ref format is 'ref: refs/remotes/origin/main'
+        with open(head) as ref:
+            return ref.read().strip().split('/')[-1]
+
+    # Get the default branch if '.git' is a file
+    command = Command("cat", f"{dot_git}")
+    result = run_command(command=command, cwd=repository, logger=logger)
+    if result.stdout is None:
+        return None
+    *_, worktree_path = result.stdout.strip().split()
+    if '/.git/' not in worktree_path:
+        return None
+    git_parent_path = re.sub(r'/?\.git/.*', '', worktree_path)
+    command = Command("git", "rev-parse", "--abbrev-ref", "HEAD")
```

Comment:
So here we get the current branch in the parent git repository. But we need to detect the default branch. Shouldn't we rather first check if the `.git` is a file and find out the real git repository and then always do the default branch checking part which is under `if dot_git.is_dir`?

Reply:
Used `git rev-parse --path-format=absolute --git-common-dir` :-)

Reply:
I don't understand what's the point here. That command only detects path to the original git repository from which the worktree has been created. But my main comment has not been addressed at all.

Can I have a question? When you do a change, do you try that the thing actually does what you expected? I guess in this case not, otherwise you would be surprised what your last change produces:

```python
>>> tmt.utils.default_branch(tmt.utils.Path('/tmp/worktree/REPRO/TREE'), tmt.log.Logger.create())
'/tmp/worktree/REPRO/.git'
```

I would recommend to write the test coverage in a way that really exercises the use case. That is in the original repository (from which the worktree is created) checkout to a different branch than the default branch `main` and then verify from the worktree that the default branch is correctly detected.

Reply:
> So here we get the current branch in the parent git repository. But we need to detect the default branch. Shouldn't we rather first check if the .git is a file and find out the real git repository and then always do the default branch checking part which is under if dot_git.is_dir? 

> **+** I don't understand what's the point here. That command only detects path to the original git repository from which the worktree has been created. But my main comment has not been addressed at all.

Hi @psss, I'm pretty confused right here. Let me paste the whole function and explain what I understand first.

```python
  2637	def default_branch(repository: Path, logger: tmt.log.Logger,
  2638	                   remote: str = 'origin') -> Optional[str]:
  2639	    """ Detect default branch from given local git repository """
  2640	    # Make sure '.git' (which is a file or a directory) is present
  2641	    dot_git = repository / '.git'
  2642	    if not dot_git.exists():
  2643	        return None
  2644	
  2645	    if not dot_git.is_file() and not dot_git.is_dir():
  2646	        return None
  2647	
  2648	    if dot_git.is_dir():
  2649	        # Make sure the '.git/refs/remotes/{remote}' directory is present
  2650	        git_remotes_dir = repository / f'.git/refs/remotes/{remote}'
  2651	        if not git_remotes_dir.exists():
  2652	            return None
  2653	
  2654	        head = repository / f'.git/refs/remotes/{remote}/HEAD'
  2655	        # Make sure the HEAD reference is available
  2656	        if not head.exists():
  2657	            subprocess.run(
  2658	                f'git remote set-head {remote} --auto'.split(), cwd=repository)
  2659	        # The ref format is 'ref: refs/remotes/origin/main'
  2660	        with open(head) as ref:
  2661	            return ref.read().strip().split('/')[-1]
  2662	
  2663	    # Get the default branch if '.git' is a file
  2664	    worktree_path = dot_git.read_text().strip().split()[-1]
  2665	    if '/.git/' not in worktree_path:
  2666	        return None
  2667	    git_parent_path = re.sub(r'/?\.git/.*', '', worktree_path)
  2668	    command = Command("git", "rev-parse", "--path-format=absolute", "--git-common-dir")
  2669	    result = run_command(command=command, cwd=Path(git_parent_path), logger=logger)
  2670	    return result.stdout.strip() if result.stdout is not None else None
  2671	
```

1. From `L2639` to `L2647` - we try to make sure `.git` is either a **file** or a **dir**;
2. From `L2648` to `L2662` - the code path to corp with `.git` which is a `dir`;
3. From `L2663` to `L2671` - the code path to corp with `.git` which is a `file`, which is based on the user case in the following:
   ```bash
    (tmt) huanli@kvm-01-guest09:foo$ git clone https://github.com/teemtee/tmt REPRO
    (tmt) huanli@kvm-01-guest09:foo$ cd REPRO
    (tmt) huanli@kvm-01-guest09:REPRO$ git worktree add TREE quay
    (tmt) huanli@kvm-01-guest09:REPRO$ cd TREE
    (tmt) huanli@kvm-01-guest09:TREE$ tmt plan show -vvv
    ...<snip>...........................................
    
    /plans/sanity/lint
                     summary Metadata used by tmt itself are valid
                    discover 
                       tests /lint/tests and /lint/plans
    ...<snip>...........................................
                     sources /tmp/foo/REPRO/TREE/plans/main.fmf
                             /tmp/foo/REPRO/TREE/plans/sanity/lint.fmf
                      fmf-id url: https://github.com/teemtee/tmt
                             ref: quay
                             name: /plans/sanity/lint
                         web https://github.com/teemtee/tmt/tree/quay/plans/sanity/lint.fmf
   ```

**`ref: quay`** is what we want to have because `quay` is not the default branch. Where is my understanding wrong? @lukaszachy，would you please help to clarify? (I'm confused) Thanks!

Reply:
> Can I have a question? When you do a change, do you try that the thing actually does what you expected? I guess in this case not, otherwise you would be surprised what your last change produces:
```python
>>> tmt.utils.default_branch(tmt.utils.Path('/tmp/worktree/REPRO/TREE'), tmt.log.Logger.create())
'/tmp/worktree/REPRO/.git'
```

Hi @psss, I don't know how to use `tmt.utils.XXX` to verify the fix via the interactive terminal of Python. Would you please share the howto which looks cool, thanks!


BTW, currently I use two ways to verify the fix of mine after updating the code.
* 01 - run the test case added
* 02 - debug the related code updated by me, e.g.
  ```python
     1	(tmt) huanli@kvm-01-guest09:TREE$ pwd
     2	/home/huanli/dev/1861/foo/REPRO/TREE
     3	(tmt) huanli@kvm-01-guest09:TREE$ git branch
     4	+ main
     5	* quay
     6	(tmt) huanli@kvm-01-guest09:TREE$ cat `pwd`/.git
     7	gitdir: /home/huanli/dev/1861/foo/REPRO/.git/worktrees/TREE
     8	(tmt) huanli@kvm-01-guest09:TREE$ python3 -m pdb ~/.virtualenvs/tmt/bin/tmt plans  show -vvv
     9	...<snip>....................................................
    10	(Pdb) b tmt/utils.py:2664
    11	Breakpoint 1 at /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py:2664
    12	(Pdb) c
    13	/plans/features/advanced
    14	...<snip>....................................................
    15	> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py(2664)default_branch()
    16	-> worktree_path = dot_git.read_text().strip().split()[-1]
    17	(Pdb) l
    18	2659 	        # The ref format is 'ref: refs/remotes/origin/main'
    19	2660 	        with open(head) as ref:
    20	2661 	            return ref.read().strip().split('/')[-1]
    21	2662 	
    22	2663 	    # Get the default branch if '.git' is a file
    23	2664B->	    worktree_path = dot_git.read_text().strip().split()[-1]
    24	2665 	    if '/.git/' not in worktree_path:
    25	2666 	        return None
    26	2667 	    git_parent_path = re.sub(r'/?\.git/.*', '', worktree_path)
    27	2668 	    command = Command("git", "rev-parse", "--path-format=absolute", "--git-common-dir")
    28	2669 	    result = run_command(command=command, cwd=Path(git_parent_path), logger=logger)
    29	(Pdb) n
    30	> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py(2665)default_branch()
    31	-> if '/.git/' not in worktree_path:
    32	(Pdb) p worktree_path
    33	'/home/huanli/dev/1861/foo/REPRO/.git/worktrees/TREE'
    34	(Pdb) b 2669
    35	Breakpoint 2 at /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py:2669
    36	(Pdb) c
    37	> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py(2669)default_branch()
    38	-> result = run_command(command=command, cwd=Path(git_parent_path), logger=logger)
    39	(Pdb) n
    40	> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py(2670)default_branch()
    41	-> return result.stdout.strip() if result.stdout is not None else None
    42	(Pdb) 
    43	--Return--
    44	> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py(2670)default_branch()->'/home/huanli...oo/REPRO/.git'
    45	-> return result.stdout.strip() if result.stdout is not None else None
    46	(Pdb) 
    47	> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py(2497)fmf_id()
    48	-> if def_branch is None:
    49	(Pdb) p def_branch
    50	'/home/huanli/dev/1861/foo/REPRO/.git'
    51	# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ I didn't realized it's not the 'main' branch :-(
    52	(Pdb) n
    53	> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py(2500)fmf_id()
    54	-> ref = run(Command("git", "rev-parse", "--abbrev-ref", "HEAD"))
    55	(Pdb) 
    56	> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py(2501)fmf_id()
    57	-> if ref != def_branch or always_get_ref:
    58	(Pdb) p ref
    59	'quay'
    60	# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Well,it's what we want to have !!!
  ```

@psss, I didn't  realized that `/home/huanli/dev/1861/foo/REPRO/.git` is the **main branch** that you mean.  Because it's not `None`, and I think it's okay to enter into the code path:
```python
2496 	    def_branch = default_branch(git_root, logger)
2497 	    if def_branch is None:
2498 	        fmf_id.ref = None
2499 	    else:
2500 	        ref = run(Command("git", "rev-parse", "--abbrev-ref", "HEAD"))
2501 	        if ref != def_branch or always_get_ref:
2502 	            fmf_id.ref = ref
```

In addition, from the output of `git branch` in the following, the worktree is `quay`, and its main branch should be `main`, right?
```
(tmt) huanli@kvm-01-guest09:TREE$ git branch
+ main
* quay
```

Please help to clarify, I'm confused that what the main branch should be if `.git` is a file, thanks!

Reply:
@psss, I updated the patch again and verified it via pdb. Please ignore the previous two comments. And I verified the patch via pdb. For details, please refer to the [output](https://github.com/idorax/tmt-dev-sandbox/blob/dev.20230420.1861/1861/pr1891/debug.md).

Reply:
> Hi @psss, I don't know how to use `tmt.utils.XXX` to verify the fix via the interactive terminal of Python. Would you please share the howto which looks that it can save a lot of time when fixing bugs? thanks!

No special magic was involved, I just called the function:

```
> python
Python 3.11.2 (main, Feb  8 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tmt
>>> tmt.utils.default_branch(tmt.utils.Path('/tmp/worktree/REPRO/TREE'), tmt.log.Logger.create())
'current-branch'
```